src/java.desktop/share/classes/com/sun/java/swing/SwingUtilities3.java

```diff
@@ -135,4 +142,66 @@ public static RepaintManager getDelegateRepaintManager(Component
         }
         return delegate;
     }
+
+    @FunctionalInterface
+    public interface UnscaledBorderPainter {
+        void paintUnscaledBorder(Component c, Graphics g,
+                                 int w, int h, double scaleFactor);
+    }
+
+    public static void paintBorder(Component c, Graphics g, int x, int y,
+                                   int w, int h, UnscaledBorderPainter painter) {
+
+        // Step 1: Reset Transform
+        AffineTransform at = null;
+        Stroke oldStk = null;
+        boolean resetTransform = false;
+        double scaleFactor = 1;
+
+        int xtranslation = x;
+        int ytranslation = y;
+        int width = w;
+        int height = h;
+
+        if (g instanceof Graphics2D) {
+            Graphics2D g2d = (Graphics2D) g;
+            at = g2d.getTransform();
+            oldStk = g2d.getStroke();
+            scaleFactor = Math.min(at.getScaleX(), at.getScaleY());
+
+            // if m01 or m10 is non-zero, then there is a rotation or shear
+            // or if scale=1, skip resetting the transform
```

Comment:
```suggestion
            // if m01 or m10 is non-zero, then there is a rotation or shear,
            // or if scale=1, skip resetting the transform in these cases
```
The added comma separates the two sentences. The added “in these cases” clarifies that we don't reset in either of the two cases: m01 or m10 is non-zero or scale is 1.